common/dynamicconfig/basic_client.go

```diff
@@ -91,105 +92,87 @@ func (bc *basicClient) GetValueWithFilters(
 func (bc *basicClient) GetIntValue(
 	name Key,
 	filters []map[Filter]interface{},
-	defaultValue int,
+	defaultValue any,
 ) (int, error) {
 	val, err := bc.getValueWithFilters(name, filters, defaultValue)
-	if err != nil {
-		return defaultValue, err
-	}
-
 	if intVal, ok := val.(int); ok {
-		return intVal, nil
+		return intVal, err
 	}
-	return defaultValue, errors.New("value type is not int")
+	intVal, _ := defaultValue.(int)
+	return intVal, errors.New("value type is not int")
```

Comment:
this does not look right, it always return error.

Reply:
same for other types

Reply:
see line 98. we only get to this point if we have an error (wrong type). the problem is we can't just return `defaultValue` since it's now an `any`. we have to convert it to an int. but what if that fails? well, who cares, we'll just return the zero value.

the fully type-safe alternative is to leave `defaultValue` as a typed param and then add a new `constrainedDefaultValues` to every call. that is cleaner in some ways, but also semantically kind of muddy. and it would make a _much_ larger diff (adding a param to all dynamic config calls).

Reply:
I mean if line 101 works, this method still returns non-nil error.

Reply:
as it should: the value read from dynamic config is not an int. the _default_ value (passed in by the caller) may or may not be an int. actually it should always be an int, unless the caller is using the new structured defaults, in which case one of the defaults should be an unconstrained value that is an int. if _that_ value is not an int, then we'll also get here, and end up with zero for the final return value. in all of those cases, this error is correct.

also note that errors from all of the Client functions are [throttle-]logged and thrown away in the Collection. they don't affect runtime behavior (unless a caller is using Client directly instead of Collection, which no one does).

Reply:
What is the purpose of line 101 then? Should remove line 101, and return 0, errors.New(...).

Reply:
discussed in person. the confusing part is that this doesn't follow the convention of "value is ignored if error is not nil". the caller (Collection) always uses the returned value, and also logs the "error". it's not really an error, it's just a statement about whether/how it had to fall back to the default.